### DISCUSSION TITLE
The design misalignments between Rspack and webpack

### DISCUSSION THREAD

**Discussion by hardfist** [web-infra-dev/rspack#5442] [Core]
This discussion channel is used to track the **design**(not implmentation) misalignments between Rspack and webpack, we need to align with webpack unless we have strong solid reason to diff with webpack.

**Comment by hardfist** [comment]
defaultValue of `optimization.removeAvailableModules` is set `true` in Rspack  https://github.com/web-infra-dev/rspack/blob/main/packages/rspack/src/config/defaults.ts#L712  but set `false` in webpack.

**Reply by JSerFeng** [reply]
Webpack can remove available duplicate modules, it achieve this using [code splitting](https://github.com/webpack/webpack/blob/main/lib/buildChunkGraph.js#L591-L599), and [RemoveParentModules](https://github.com/webpack/webpack/blob/main/lib/optimize/RemoveParentModulesPlugin.js), the `optimization.removeAvailableModules` config only disable the latter, so the optimization from code splitting, it's still enabled.
So for Webpack whether disable `optimization.removeAvailableModules` or not, available duplicate modules will be removed eventually, so Rspack enables it by default, and if you don't want this optimization, you can turn it off.

**Comment by LingyuCoder** [comment]
`compilation.hooks.runtimeModule` is different with webpack.

In Webpack, plugin can get the module instance directly through `compilation.hooks.runtimeModule`, and then modify its content by  overriding its `getGenerateCode()` method like below:

```js
compilation.hooks.runtimeModule.tap('xxx', (module, chunk) => {
    if (chunk.name === 'xxx' && (module.name === 'xxx' || module.constructor.name === 'xxx')) {
        const originSource = module.getGeneratedCode();
        const appendSource = "/* new source */";
        module.getGeneratedCode = () => `${originSource}\n${appendSource}`;
    }
});
```

But in Rspack, since the injection of the runtime module is on the Rust side, and it is very hard to port the entire runtime module + runtime requirements to the JavaScript side.
So it is necessary to convert the Rust-side runtime module into a JsModule, transmit it to the JavaScript side for modification, and then send it back to the Rust side.
Therefore, the modification of the runtime module is based on string transmission, and it cannot be as flexible as Webpack, only supporting the replacement of module source:

```js
compilation.hooks.runtimeModule.tap("TestFakePlugin", (module, chunk) => {
    if (chunk.name === 'xxx' && (module.name === 'xxx' || module.constructorName === 'xxx')) {
        const originSource = module.source.source.toString('utf-8');
        const appendSource = "blabla";
        module.source.source = Buffer.from(
            `${originSource}\n${appendSource}`,
            "utf-8"
        );
    }
});
```
